middleware/2fa.js

```diff
@@ -0,0 +1,287 @@
+    const keyStore = {
+        async getKey() {
+            return nearAPI.KeyPair.fromString('bs');
+        },
+    };
+    // check account code_hash
+    const near = await nearAPI.connect({
+        deps: { keyStore },
+        masterAccount: creatorKeyJson && creatorKeyJson.account_id,
+        nodeUrl: process.env.NODE_URL
+    });
+    const nearAccount = new nearAPI.Account(near.connection, accountId);
+    const state = await nearAccount.state();
+    console.log(state);
+    return state.code_hash !== '11111111111111111111111111111111';
+};
+
+/********************************
+@warn protect these routes using checkAccountOwnership middleware from app.js
+@warn Requires refactor
+********************************/
+// http post http://localhost:3000/2fa/getAccessKey accountId=mattlock
+// http post https://helper.testnet.near.org/2fa/getAccessKey accountId=mattlock
+// http post https://near-contract-helper-2fa.onrender.com/2fa/getAccessKey accountId=mattlock
+// Call this to get the public key of the access key that contract-helper will be using to confirm multisig requests
+const getAccessKey = async (ctx) => {
+    const { accountId } = ctx.request.body;
+    const { publicKey } = await getDetermKey(accountId);
+    ctx.body = {
+        success: true,
+        publicKey
+    };
+};
+// https://near-contract-helper-2fa.onrender.com/
+// http post http://localhost:3000/2fa/init accountId=mattlock method:='{"kind":"2fa-email","detail":"matt@near.org"}'
+// http post https://near-contract-helper-2fa.onrender.com/2fa/init accountId=mattlock method:='{"kind":"2fa-email","detail":"matt@near.org"}'
+// Call ONCE to enable 2fa on this account. Adds a twoFactorMethod (passed in body) where kind should start with '2fa-'
+// This WILL send the initial code to the method specified ['2fa-email', '2fa-phone']
+const initCode = async (ctx) => {
+    const { accountId, method, testing, code_hash } = ctx.request.body;
+    const [account] = await models.Account.findOrCreate({ where: { accountId } });
+    if (!account) {
+        ctx.throw(401, 'account should be created');
+        return;
+    }
+    if (!method || !method.kind || !method.detail) {
+        ctx.throw(401, 'method arguments invalid');
+        return;
+    }
+    const { kind, detail } = method;
+    // validate method kind
+    if (!twoFactorMethods.includes(kind)) {
+        ctx.throw(401, 'invalid 2fa method ' + kind);
+        return;
+    }
+    const hasContractDeployed = await isContractDeployed(accountId);
+    // check recover methods
+    let [twoFactorMethod] = await account.getRecoveryMethods({ where: {
+        kind: {
+            [Op.startsWith]: '2fa-'
+        },
+    }});
+    // check if multisig contract is already deployed
+    if (hasContractDeployed || !!code_hash) {
+        // check to see if they already have at least 1 twoFactorMethod
+        if (twoFactorMethod) {
+            ctx.throw(401, 'account with multisig contract already has 2fa method');
```

Comment:
hmm, so does it mean that 2FA method cannot be changed later?

Reply:
we can provide `2fa/update` endpoint, but out of scope for phase 0? @kcole16